src/rust/engine/graph/src/lib.rs

```diff
@@ -676,32 +696,59 @@ impl<N: Node> Graph<N> {
   }
 
   ///
-  /// Gets the generations of the dependencies of the given EntryId, (re)computing or cleaning
-  /// them first if necessary.
+  /// Compares the generations of the dependencies of the given EntryId to their previous
+  /// generation values (re-computing or cleaning them first if necessary), and returns true if any
+  /// dependency has changed.
   ///
-  async fn dep_generations(
+  async fn dependencies_changed(
     &self,
     entry_id: EntryId,
+    previous_dep_generations: Vec<Generation>,
     context: &N::Context,
-  ) -> Result<Vec<Generation>, N::Error> {
-    let generations = {
+  ) -> bool {
+    let generation_matches = {
       let inner = self.inner.lock();
-      inner
+      let dependency_ids = inner
         .pg
         .neighbors_directed(entry_id, Direction::Outgoing)
-        .map(|dep_id| {
+        .collect::<Vec<_>>();
+
+      if dependency_ids.len() != previous_dep_generations.len() {
+        // If we don't have the same number of current dependencies as there were generations
+        // previously, then they cannot match.
+        return true;
```

Comment:
It was a bit jarring reading this early return. Is there any reason not to lift the `dependency_ids` collection and length check with early return out of the generation_matches binding block?

Reply:
Mostly that we'd need to re-acquire the lock to get access to the `Entry`s again after the check.